modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/navigate.ts

```diff
@@ -16,34 +16,32 @@
  * Performs navigation to the given url. If SPA is enabled, it will route the
  * request through the SPA engine. If not, it will simple change the document
  * location.
- * @param {string | URL} url Destination URL to navigate
- * @param {?object} listeners Object with key-value pairs with callbacks for
- * specific page lifecycle events
- * @review
  */
 
-export default function (url, listeners) {
-	let urlString = url;
+import * as SPATypes from 'frontend-js-spa-web/src/main/resources/META-INF/resources/init';
+import * as StateTypes from '@liferay/frontend-js-state-web/src/main/resources/META-INF/resources/State';
 
-	if (url?.constructor?.name === 'URL') {
-		urlString = String(url);
-	}
+export default function (
+	url: string | URL,
+	listeners?: {[key: string]: Callback}
+) {
+	url = url instanceof URL ? String(url) : url;
```

Comment:
We do a lot of type narrowing that TS can't pick up like in here. In this case, `instanceof` sounds like a better approach overall, but wondering what we do in general.

We're likely going to need to rewrite code because this code style is not easily expressed with types (that I could figure). There are other ways to guard types like user-defined or creating intermediate helper methods like `getURLString(url: string | URL): string` that help TS here. 

Wondering how this will shape out in some of the more complex isomorphisms we have 😈

Reply:
I have a similar in [FilePath](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts) I think...

I use the [AnyPath](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts#L9) type for `string | FilePath` and [FilePath.coerce](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts#L14) to do the cast.